src/_pytest/main.py

```diff
@@ -518,6 +518,28 @@ def __repr__(self) -> str:
             self.testscollected,
         )
 
+    @property
+    def shouldstop(self) -> Union[bool, str]:
+        return self._shouldstop
+
+    @shouldstop.setter
+    def shouldstop(self, value: Union[bool, str]) -> None:
+        """Prevent plugins from setting this to False once it has been set."""
+        if value is False and self._shouldstop:
+            return
```

Comment:
I think silently ignoring is no good, it will be confusing why there is no effect.

The two options are issuing a warning, or raising an exception. Since we're already stopping raising an exception seems kinda pointless, so I think we should go with a warning. I think it can be something like this:

```py
# The runner checks shouldfail and assumes that if it is set we are definitely stopping,
# so prevent unsetting it.
warnings.warn(PytestWarning("session.shouldfail cannot be unset after it has been set; ignoring."), stacklevel=2)
```

Unfortunately it does mean you'll need to write extra tests to cover these warnings, let me know if you need help with that.

Reply:
Okay sounds good, I've added the warnings and and a unit test for the `shouldfail` scenario called `test_shouldfail_is_sticky`. I'm using `pytest_sessionfinish` to trigger the bad behavior. However this same approach did not work for `shouldstop` - it seems like `shouldstop` was still `False` in `pytest_sessionfinish` even if I changed `pytest.fail` to `pytest.exit`. Any ideas on how to test the `shouldstop` scenario?

Reply:
The only thing which sets `shouldstop` in pytest core is `--stepwise`, which actually has the same problem as `--maxfail`. Since I've already checked out the branch, I've pushed the new test along with some tweaks. This is now good to go, thanks!